Caltech-IPAC/jupyter_firefly_extensions
# DM-16021-16022-14659: extensions working and widget combined

### Major new features:

 - We are now using full extensions (not a single mime extension)
 - There are three extension:
   1. file extension - it you double click on a fits file it show in a tab.
   2. Slate extension - 2 parts:
         1. you and click on the open firefly command a see a new slate window.
         1. firefly_client will open a slate tab for you.
   3. Widget extension 
         1. widgets parts port from firefly_widgets
         1. new widget - slate widget - has a function that will return a FireflyClient instance to control it.
- All works with one instance of firefly
   - extreme example- 
         1. double click on a fits file.
         2. using the command open a slate tab and use UI to add stuff
         3. load a notebook that has a FireflyClient, it will open another tab that you control with the API
         4. load the a image widget with another notebook
         5. load a SlateWidget and control it in the notebook with a FireflyClient.

  

### _To test:_

- build the related firefly pull request, and checkout the related firefly_client pull request
- make a directory with some fit files and the file from `jupyter_firefly_extensions/examples`
- follow the instruction in the read me to make sure you have the extension installed
- set you firefly instance as the readme describes
- note new step: `jupyter labextension install @jupyter-widgets/jupyterlab-manager`
- start jupyter lab
- _Then:_
     - Double click on a fits file - see a new tab with the image
     - load the notebook- `slate-demo-explicit.ipynb`, go though all the cells
     - click on the firefly command, load some stuff 
     - 	load the notebook-  `slate-widget-demo.ipnb`, go thought all the cells
     - load 	the notebook `Images and Tables.ipynb`,  go thought all the cells



### _Known bugs:_

 - I think there are many that we might have to work on in other tickets.
 - ~~For the slate widget, it does not always refresh when you tab to it.~~, _fixed in last commit_
 - tab names need improvement
 - the tables do not have a separate "viewer group" so it is possible for them to show up at unexpected times. 

### _Not yet implemented:_

- [ ] For fits viewer extension- direct file load, not though the browser
- [ ] make widget produce a screen shot
- [ ] make slate widget have option  to move to slate view tab
- [ ] Add firefly to the launcher window
- [ ] Make file analysis work for the fits file extension
